app/javascript/mastodon/db/modifier.js

```diff
@@ -0,0 +1,86 @@
+import asyncDB from './async';
+
+const limit = 1024;
+
+function put(name, objects, callback) {
+  asyncDB.then(db => {
+    const store = db.transaction(name, 'readwrite').objectStore(name);
+    const index = store.index('id');
+    let left = objects.length;
+
+    function done() {
+      left--;
+
+      if (left <= 0) {
+        store.count().onsuccess = count => {
+          const excess = count.target.result - limit;
+
+          if (excess > 0) {
+            store.getAll(null, excess).onsuccess =
+              retrieval => callback(retrieval.target.result.map(({ id }) => id));
+          }
+        };
+      }
+    }
+
+    objects.forEach(object => {
+      function add() {
+        store.add(object).onsuccess = done;
+      }
+
+      index.getKey(object.id).onsuccess = retrieval => {
+        if (retrieval.target.result) {
+          store.delete(retrieval.target.result).onsuccess = add;
+        } else {
+          add();
+        }
```

Comment:
Instead of doing `getKey` to check for existence, then `delete()` if it exists, then call `add()` to re-insert it, you can just use the `put()` API which does the same thing. `put()` will overwrite any existing values. It also doesn't throw if the value already exists.

Reply:
Actually this `put` method was named from IndexedDB's `put`. However, it is different from `put` in terms that it always creates a new record. Thanks to that, a newly fetched record always has a new key. It allows to find out-dated records by sorting the key, corresponding to the fetched time.